internal/pkg/earpointer/analysis.go

```diff
@@ -0,0 +1,649 @@
+	switch op.Op {
+	case token.ARROW: // channel read
+		// Use array read to simulate channel read.
+		vis.processHeapAccess(op, op.X, anyIndexField)
+	case token.MUL:
+		// v = *addr
+		// The LHS is a register; the RHS can be a (memory) address, a global
+		// variable, or a free variable.
+		// t1 = *addr is implemented as "addr[directPointToField] = t1",
+		// i.e., address addr points to value t1.
+		vis.processAddressToValue(op.X, op)
+	}
+}
+
+func (vis *visitor) visitBinOp(op *ssa.BinOp) {
+	// Numeric binop will be skipped.
+	// The only binop that matters now is "+" that concatenates strings.
+	if op.Op == token.ADD {
+		vis.unifyLocals(op, op.X)
+		vis.unifyLocals(op, op.Y)
+	}
+}
+
+func (vis *visitor) visitTypeAssert(assert *ssa.TypeAssert) {
+	if !assert.CommaOk {
+		// t1 = typeassert t0.(*int)
+		vis.unifyLocals(assert, assert.X)
+	}
+	// The comma case will be processed by the subsequent extract instruction.
+}
+
+func (vis *visitor) visitExtract(extract *ssa.Extract) {
+	switch base := extract.Tuple.(type) {
+	case *ssa.TypeAssert:
+		// t1 = typeassert,ok t0.(*int)
+		// t2 = extract t1 0
+		if base.CommaOk && extract.Index == 0 {
+			vis.unifyLocals(extract, base.X)
+		}
+	case *ssa.Next:
+		// t1 = range t0
+		// t2 = next t1   // return (ok bool, k Key, v Value)
+		// t3 = extract t2 ?
+		if rng, ok := base.Iter.(*ssa.Range); ok {
+			if mayShareObject(extract) && extract.Index == 2 { // extract the value v
+				// The value from the map t0 flows into the field: t0[*] = v.
+				vis.processHeapAccess(extract, rng.X, anyIndexField)
+			}
+		}
+	case *ssa.Lookup:
+		// t2 = t0[t1],ok
+		// t3 = extract t2 ?
+		if base.CommaOk && extract.Index == 0 {
+			vis.processMemoryIndexAccess(extract, base.X, base.Index)
+		}
+	default: // Other cases: model "r1 = extract r0 #n" as "r0["n"] = r1".
+		field := Field{Name: strconv.Itoa(extract.Index)}
+		vis.processHeapAccess(extract, base, field)
+	}
+}
+
+func (vis *visitor) visitNext(next *ssa.Next) {
+	// The next instruction will be processed by the subsequent extract
+	// instruction.
+}
+
+func (vis *visitor) visitSlice(slice *ssa.Slice) {
+	vis.unifyLocals(slice, slice.X)
+}
+
+func (vis *visitor) visitMakeInterface(makeInterface *ssa.MakeInterface) {
+	vis.unifyLocals(makeInterface, makeInterface.X)
+}
+
+func (vis *visitor) visitSend(send *ssa.Send) {
+	vis.processHeapAccess(send.X, send.Chan, anyIndexField)
+}
+
+func (vis *visitor) visitRange(rng *ssa.Range) {
+	// t1 = range t0: unify t0 and t1.
+	// It will be processed by the subsequent extract instruction that also handles the relevant Next.
+}
+
+func (vis *visitor) VisitSelect(s *ssa.Select) {
+	// TODO: to be implemented.
+}
+
+func (vis *visitor) visitCall(call ssa.CallCommon, instr ssa.Instruction) {
+	// TODO: to be added.
+}
+
+// Process a load/store using an index (which can be constant).
+func (vis *visitor) processMemoryIndexAccess(data ssa.Value, base ssa.Value, index ssa.Value) {
```

Comment:
Nit: It might be easy to think that `IndexAccess` doesn't include a map key.  The spec refers to them as [Index Expressions](https://golang.org/ref/spec#Index_expressions).  So maybe `processIndexExpression` would be more explicit here?

I'm terrible with naming, though.  If anyone has a real opinion, I'd defer to that.